jax/_src/third_party/scipy/special.py

```diff
@@ -0,0 +1,293 @@
+      +8.68029542941784300606e-4,
+      +4.12142090722199792936e-2,
+      +1.00000000000000000118e0,
+    ])
+
+    fresnl_fn = jnp.array([
+      +4.21543555043677546506e-1,
+      +1.43407919780758885261e-1,
+      +1.15220955073585758835e-2,
+      +3.45017939782574027900e-4,
+      +4.63613749287867322088e-6,
+      +3.05568983790257605827e-8,
+      +1.02304514164907233465e-10,
+      +1.72010743268161828879e-13,
+      +1.34283276233062758925e-16,
+      +3.76329711269987889006e-20,
+    ])
+
+    fresnl_fd = jnp.array([
+      +1.00000000000000000000e0,
+      +7.51586398353378947175e-1,
+      +1.16888925859191382142e-1,
+      +6.44051526508858611005e-3,
+      +1.55934409164153020873e-4,
+      +1.84627567348930545870e-6,
+      +1.12699224763999035261e-8,
+      +3.60140029589371370404e-11,
+      +5.88754533621578410010e-14,
+      +4.52001434074129701496e-17,
+      +1.25443237090011264384e-20,
+    ])
+
+    fresnl_gn = jnp.array([
+      +5.04442073643383265887e-1,
+      +1.97102833525523411709e-1,
+      +1.87648584092575249293e-2,
+      +6.84079380915393090172e-4,
+      +1.15138826111884280931e-5,
+      +9.82852443688422223854e-8,
+      +4.45344415861750144738e-10,
+      +1.08268041139020870318e-12,
+      +1.37555460633261799868e-15,
+      +8.36354435630677421531e-19,
+      +1.86958710162783235106e-22,
+    ])
+
+    fresnl_gd = jnp.array([
+      +1.00000000000000000000e0,
+      +1.47495759925128324529e0,
+      +3.37748989120019970451e-1,
+      +2.53603741420338795122e-2,
+      +8.14679107184306179049e-4,
+      +1.27545075667729118702e-5,
+      +1.04314589657571990585e-7,
+      +4.60680728146520428211e-10,
+      +1.10273215066240270757e-12,
+      +1.38796531259578871258e-15,
+      +8.39158816283118707363e-19,
+      +1.86958710162783236342e-22,
+    ])
+
+  x = jnp.abs(xxa)
+
+  x2 = x * x
+
+  # Infinite x values
+  s_inf = c_inf = 0.5
+
+  # Small x values
+  t = x2 * x2
+
+  if single_precision:
+    s_small = x * x2 * jnp.polyval(fresnl_sn, t)
+    c_small = x * jnp.polyval(fresnl_cn, t)
+  else:
+    s_small = x * x2 * jnp.polyval(fresnl_sn[:6], t) / jnp.polyval(fresnl_sd[:7], t)
+    c_small = x * jnp.polyval(fresnl_cn[:6], t) / jnp.polyval(fresnl_cd[:7], t)
+
+  # Large x values
+
+  sinpi, cospi = sincospisquaredhalf(x)
+
+  if single_precision:
+    c_large = c_inf
+    s_large = s_inf
+  else:
+    c_large = 0.5 + 1 / (jnp.pi * x) * sinpi
+    s_large = 0.5 - 1 / (jnp.pi * x) * cospi
+
+  # Other x values
+  t = jnp.pi * x2
+  u = 1.0 / (t * t)
+  t = 1.0 / t
+
+  if single_precision:
+    f = 1.0 - u * jnp.polyval(fresnl_fn, u)
+    g = t * jnp.polyval(fresnl_gn, u)
+  else:
+    f = 1.0 - u * jnp.polyval(fresnl_fn, u) / jnp.polyval(fresnl_fd, u)
+    g = t * jnp.polyval(fresnl_gn, u) / jnp.polyval(fresnl_gd, u)
+
+  t = jnp.pi * x
+  c_other = 0.5 + (f * sinpi - g * cospi) / t
+  s_other = 0.5 - (f * cospi + g * sinpi) / t
+
+  isinf = jnp.isinf(xxa)
+  small = x2 < 2.5625
+  large = x > 36974.0
```

Comment:
This is what is confusing to me: we have explicit logic for `x > 36974`, but you mentioned the docs say the outputs are only good for $|x| < 10$. How do I reconcile those two things?

Reply:
Well I did not claim (nor they did) that it would fail for large values, but reading the source code, this is what they say

>  * ACCURACY:
>  *
>  *  Relative error.
>  *
>  * Arithmetic  function   domain     # trials      peak         rms
>  *   IEEE       S(x)      0, 10       30000       1.1e-6      1.9e-7
>  *   IEEE       C(x)      0, 10       30000       1.1e-6      2.0e-7
>  */

for the single-precision version, and

>  * ACCURACY:
>  *
>  *  Relative error.
>  *
>  * Arithmetic  function   domain     # trials      peak         rms
>  *   IEEE       S(x)      0, 10       10000       2.0e-15     3.2e-16
>  *   IEEE       C(x)      0, 10       10000       1.8e-15     3.3e-16
>  *   DEC        S(x)      0, 10        6000       2.2e-16     3.9e-17
>  *   DEC        C(x)      0, 10        5000       2.3e-16     3.9e-17
>  */

for the double-precision version.

So while they have specific cases for large values, I don't know what one can expect regarding their accuracy. I could not find more documentation about them than that of https://www.netlib.org/cephes/.

Reply:
OK, I guess I misunderstood your earlier comment in that case.

All the details aside, what I'm saying is this: we should document and test the range of values over which we expect this function to be accurate.

Reply:
I added some comment about this, but I cannot find a range that works (where previously I swear it would), but I cannot spend more time on this PR at the moment :/

Reply:
There's still some work to be done to land this – are you saying you no longer want to contribute this?

Reply:
Not that I don't want to contribute to this, but I will need a pause on this for at least one week, and continue later (probably after 2nd of September). If anyone is willing to give some help, I'd gladly take some :-)

Reply:
Sure, sounds good!

Reply:
Hello @jakevdp! So I found out why tests were not working anymore, a dumb typo felt in the Cephes coefficients for single precision, but it's now fixed. I have adjusted the scale so that it now test a range we are sure the coefficients were developed for. It is documented in the docstring. For the half-float dtypes, see the other comments thread :-)